<a href="https://colab.research.google.com/github/sangita-mitra/Practice-Data-Science-Project/blob/main/Fake_News_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [55]:
#importing libraries
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [56]:
#Read the data
df=pd.read_csv('news.csv')
df.head()

Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]

In [57]:
#Check shape of data
df.shape

(6335, 4)

In [58]:
#Check data types
df.dtypes

Unnamed: 0     int64
title         object
text          object
label         object
dtype: object

In [59]:
df.describe()

Unnamed: 0
count   6335.000000
mean    5280.415627
std     3038.503953
min        2.000000
25%     2674.500000
50%     5271.000000
75%     7901.000000
max    10557.000000

In [60]:
#Get the labels of dataframe
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

Texts are labeled as Fake and Real.

In [61]:
#convert the data type of text to string
df['text']= df['text'].values.astype('str')

In [62]:
#Split the dataset into training and testing sets in 70:30 rtaio
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.3, random_state=7)

In [63]:
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(4434,) (1901,) (4434,) (1901,)


TF-IDF is an abbreviation for Term Frequency Inverse Document Frequency. This is very common algorithm to transform text into a meaningful representation of numbers which is used to fit machine algorithm for prediction. 

Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

In [64]:
#Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

#Fit and transform train set, transform test set

tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

Passive Aggresive Classifier is an online machine learning algorithms where input data comes in sequential order and the machine learning model is updated step-by-step, as opposed to batch learning, where the entire training dataset is used at once. We’ll fit this on tfidf_train and y_train. Then, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [65]:
#Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=100)
pac.fit(tfidf_train,y_train)
#Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')
print(f"Classification Report : \n\n{classification_report(y_test, y_pred)}")

Accuracy: 92.69%
Classification Report : 

              precision    recall  f1-score   support

        FAKE       0.93      0.93      0.93       974
        REAL       0.92      0.93      0.93       927

    accuracy                           0.93      1901
   macro avg       0.93      0.93      0.93      1901
weighted avg       0.93      0.93      0.93      1901



We get an accuracy of 92.69% with this model. 

In [66]:
#Build confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[904,  70],
       [ 69, 858]])

So with this model, we have 904 true positives, 858 true negatives, 69 false positives, and 70 false negatives.